### PR TITLE
CAMS-461 - Migrate consolidations from ACMS

### PR DESCRIPTION
**Pull request by btposey**
# Purpose

Import consolidated cases from ACMS into CAMS.

# Major Changes

* Added ACMS SQL Server database gateway
* Added durable functions to orchestrate the ETL between ACMS and CAMS

# Testing/Validation

* Unit tests
* Manual testing to confirm function

# Notes

* Adds ACMS_DATABASE environment variable as a new configuration item.

# Definition of Done:

- [ ] Acceptance criteria met - The system meets all acceptance criteria
- [ ] Feature toggles created - Features that are deployed but not released have toggles
- [ ] Design QA passed - All visual UI looks and works the way it’s supposed to
- [ ] Usability validated - Team ensures the working code is easy to use
- [ ] Accessibility tests passed - System passes automated and manual accessibility tests and supports applicable devices
- [ ] Code refactored for clarity - Developers can understand the work simply by reviewing the code
- [ ] Dependency Rule followed - More important code doesn’t directly depend on less important code
- [ ] Development debt eliminated - UX and code aligns to the team’s latest understanding of the domain
- [ ] Source code merged - Code has been merged into the main branch
- [ ] 90% unit test coverage achieved - Automated test coverage tools indicate coverage of >= 90%
- [ ] Code reviewed - Code is reviewed by at least two other team members before being merged
- [ ] Code quality checks passed - Code passes all automated quality checks
- [ ] Security scans passed - Code passes expected scans for vulnerabilities and compliance issues
- [ ] Threat model updated - The threat model incorporates new threats and remediations
- [ ] Build process updated - Automated build process include code supporting the user story
- [ ] Load/performance tests passed - Performance tests include the functionality of the user story
- [ ] Documents are updated - All required documentation is up to date and version controlled
- [ ] Code is deployed - Code is deployed to the highest production like environment

### CODE REVIEW

File: backend/functions/lib/adapters/gateways/acms/acms.gateway.ts [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+      value: PAGE_SIZE,
+    });
+
+    input.push({
+      name: `offset`,
+      type: mssql.Int,
+      value: PAGE_SIZE * (predicateAndPage.pageNumber - 1),
+    });
+
+    // RIGHT(CONCAT('000', CAST(CONSOLIDATED_CASE_NUMBER AS VARCHAR)), 10)
```
Comment (btposey) [inline backend/functions/lib/adapters/gateways/acms/acms.gateway.ts:95]: Remove
```suggestion
```

File: backend/functions/lib/use-cases/acms-orders/acms-orders.ts [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    try {
+      const casesRepo = Factory.getCasesRepository(context);
+      const dxtr = Factory.getCasesGateway(context);
+      const acms = Factory.getAcmsGateway(context);
+
+      const basics = await acms.getConsolidationDetails(context, leadCaseId);
+      const leadCase = await dxtr.getCaseSummary(context, leadCaseId);
+
+      // NOTE! Azure suggests that all work be IDEMPOTENT because activities run _at least once_.
+      // Check if exported child cases have already been migrated.
```
Comment (jamesobrooks) [inline backend/functions/lib/use-cases/acms-orders/acms-orders.ts:71]: For the idempotent note, I'd like to add it to a JSDOC. The other one seems like a TODO?

File: backend/functions/migration/loadConsolidations.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+df.app.orchestration(MAIN_ORCHESTRATOR, main);
+df.app.orchestration(SUB_ORCHESTRATOR_ETL, subOrchestratorETL);
+df.app.orchestration(SUB_ORCHESTRATOR_PAGING, subOrchestratorPaging);
+df.app.activity(GET_CONSOLIDATIONS, getConsolidations);
+df.app.activity(GET_PAGE_COUNT, getPageCount);
+df.app.activity(MIGRATE_CONSOLIDATION, migrateConsolidation);
+df.app.activity(FLATTEN_BOUNDING_ARRAYS, flattenBoundingArrays);
+
+app.http('dfClient', {
+  route: 'orchestrators/orchestrator',
```
Comment (jamesobrooks) [inline backend/functions/migration/loadConsolidations.ts:30]: If we end up keeping the client, we might consider changing this route.

Comment (btposey) [reply backend/functions/migration/loadConsolidations.ts:30]: Agreed.

File: backend/functions/lib/adapters/gateways/abstract-mssql-client.exception.test.ts [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+import { AbstractMssqlClient } from './abstract-mssql-client';
+import { ApplicationContext } from '../types/basic';
+
+// Setting default Jest mocks for mssql
+// eslint-disable-next-line no-var
+var connectionError = new ConnectionError('');
+// eslint-disable-next-line no-var
+var requestError = new RequestError('');
+// eslint-disable-next-line no-var
+var mssqlError = new MSSQLError('');
```
Comment (btposey) [inline backend/functions/lib/adapters/gateways/abstract-mssql-client.exception.test.ts:13]:
```suggestion
const connectionError = new ConnectionError('');
const requestError = new RequestError('');
const mssqlError = new MSSQLError('');
```

File: backend/functions/lib/use-cases/acms-orders/acms-orders.ts [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+};
+
+export type Predicate = {
+  divisionCode: string;
+  chapter: string;
+};
+
+export type PredicateAndPage = Predicate & {
+  pageNumber: number;
+};
```
Comment (btposey) [inline backend/functions/lib/use-cases/acms-orders/acms-orders.ts:24]: Maybe prefix these types with Acms*

File: backend/functions/lib/adapters/gateways/abstract-mssql-client.exception.test.ts (hunk 4 above)
Comment (btposey) [reply backend/functions/lib/adapters/gateways/abstract-mssql-client.exception.test.ts:13]: LOoks like this is needed.

File: backend/functions/migration/orchestration/orchestrator.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
+import { Bounds, Predicate } from '../../lib/use-cases/acms-orders/acms-orders';
+import { FLATTEN_BOUNDING_ARRAYS, SUB_ORCHESTRATOR_PAGING } from '../loadConsolidations';
+import { OrchestrationContext } from 'durable-functions';
+
+export function* main(context: OrchestrationContext) {
```
[the remark this answers is not in the document]
Comment (jamesobrooks) [reply backend/functions/migration/orchestration/orchestrator.ts:5]: Yes, it is a generator function which involves the yield.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/function*

**Review by jamesobrooks — DISMISSED** [review]
The API currently does not work. In addition to finishing the goals on the ensemble notes doc, we need to verify proper behavior before merging.

File: backend/functions/lib/adapters/gateways/mongo/orders.mongo.repository.ts [hunk 7]
```diff
@@ -16,10 +16,33 @@ const COLLECTION_NAME = 'orders';
...
+  public static dropInstance() {
+    if (OrdersMongoRepository.referenceCount > 0) OrdersMongoRepository.referenceCount--;
+    if (OrdersMongoRepository.referenceCount < 1) {
+      OrdersMongoRepository.instance.client.close().then();
+      OrdersMongoRepository.instance = null;
+    }
+  }
+
+  public release() {
+    console.log('***** RELEASE CALLED **********');
```
Comment (btposey) [inline backend/functions/lib/adapters/gateways/mongo/orders.mongo.repository.ts:42]:
```suggestion
```

File: backend/functions/lib/adapters/gateways/mongo/orders.mongo.repository.ts [hunk 8]
```diff
@@ -16,10 +16,33 @@ const COLLECTION_NAME = 'orders';
...
+    if (!OrdersMongoRepository.instance)
+      OrdersMongoRepository.instance = new OrdersMongoRepository(context);
+    OrdersMongoRepository.referenceCount++;
+    return OrdersMongoRepository.instance;
+  }
+
+  public static dropInstance() {
+    if (OrdersMongoRepository.referenceCount > 0) OrdersMongoRepository.referenceCount--;
+    if (OrdersMongoRepository.referenceCount < 1) {
+      OrdersMongoRepository.instance.client.close().then();
```
Comment (btposey) [inline backend/functions/lib/adapters/gateways/mongo/orders.mongo.repository.ts:36]: Never tried this out but what happens when an error is thrown and we don't have an error handler in the chain?

**Pull request closed without merging**